osquery/events/darwin/endpointsecurity.cpp

```diff
@@ -0,0 +1,220 @@
+    case ES_NEW_CLIENT_RESULT_ERR_NOT_ENTITLED:
+      return Status::failure(1, "EndpointSecurity client lacks entitlement");
+    case ES_NEW_CLIENT_RESULT_ERR_NOT_PERMITTED:
+      return Status::failure(
+          1, "EndpointSecurity client lacks user TCC permissions");
+    case ES_NEW_CLIENT_RESULT_ERR_NOT_PRIVILEGED:
+      return Status::failure(1,
+                             "EndpointSecurity client is not running as root");
+    case ES_NEW_CLIENT_RESULT_ERR_TOO_MANY_CLIENTS:
+      return Status::failure(
+          1, "Too many EndpointSecurity clients running on system");
+    default:
+      return Status::failure(1, "EndpointSecurity: Unknown Error");
+    }
+  } else {
+    return Status::failure(1,
+                           "EndpointSecurity is only available on macOS 10.15");
+  }
+}
+
+void EndpointSecurityPublisher::configure() {
+  if (es_client_ == nullptr) {
+    return;
+  }
+
+  auto cache = es_clear_cache(es_client_);
+  if (cache != ES_CLEAR_CACHE_RESULT_SUCCESS) {
+    VLOG(1) << "Couldn't clear cache for EndpointSecurity client";
+    return;
+  }
+
+  for (auto& sub : subscriptions_) {
+    auto sc = getSubscriptionContext(sub->context);
+    auto events = sc->es_event_subscriptions_;
+    auto es_sub = es_subscribe(es_client_, &events[0], events.size());
+    if (es_sub != ES_RETURN_SUCCESS) {
+      VLOG(1) << "Couldn't subscribe to ES subsystem";
+    }
+  }
+}
+
+void EndpointSecurityPublisher::tearDown() {
+  if (es_client_ == nullptr) {
+    return;
+  }
+  es_unsubscribe_all(es_client_);
+
+  // calling es_delete_client on macOS 10.15.0 - 10.15.3 leads to a crash
+  if (es_client_success_ && macos_15_4_higher_) {
+    es_delete_client(es_client_);
+  }
+}
+
+static inline std::string getPath(const es_process_t* p) {
+  return p->executable->path.length > 0 ? p->executable->path.data : "";
+}
+
+static inline std::string getSigningId(const es_process_t* p) {
+  return p->signing_id.length > 0 ? p->signing_id.data : "";
+}
+
+static inline std::string getTeamId(const es_process_t* p) {
+  return p->team_id.length > 0 ? p->team_id.data : "";
+}
+
+static inline std::string getStringFromToken(es_string_token_t* t) {
+  return t->length > 0 && t->data != nullptr ? t->data : "";
+}
+
+static inline std::string getCwdPath(pid_t pid) {
+  struct proc_vnodepathinfo vpi {};
+  auto bytes = proc_pidinfo(pid, PROC_PIDVNODEPATHINFO, 0, &vpi, sizeof(vpi));
+  return bytes <= 0 ? "" : vpi.pvi_cdir.vip_path;
+}
+
+static inline std::string getCDHash(const es_process_t* p) {
+  std::stringstream hash;
+  for (unsigned char i : p->cdhash) {
+    hash << std::hex << std::setfill('0') << std::setw(2)
+         << static_cast<unsigned int>(i);
+  }
+  auto s = hash.str();
+  return s.find_first_not_of(s.front()) == std::string::npos ? "" : s;
+}
+
+void EndpointSecurityPublisher::handleMessage(const es_message_t* message) {
+  if (message == nullptr) {
+    return;
+  }
+
```

Comment:
It would be helpful to know if ES events are being dropped (e.g. message queue full) by checking for gaps in `message->seq_num` .

Reply:
Yep, I just caught that in the WWDC EndpointSecurity session, it's on my list :)